appcelerator/titanium_mobile
# [TIMOB-24605] Android: Fixed crashes caused by loading image URLs when JS is running on main thread

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-24605
- https://jira.appcelerator.org/browse/TIMOB-26249

**Summary:**
- Fixed image URL crash if JS running on main thread for:
  * [TIMOB-24605](https://jira.appcelerator.org/browse/TIMOB-24605) `View` object's "backgroundImage" property.
  * [TIMOB-26249](https://jira.appcelerator.org/browse/TIMOB-26249) `ImageView` object's `image` property while "autorotate" is `true`.
  * `ImageView` object's "defaultImage" property.
- Modified Java "TiImageHelper" class to support reading JPEG EXIF orientation from a stream.
  * Allows `ImageView` to auto-rotate JPEG downloaded from the Internet if needed.
- Added new Java class "TiInputStreamWrapper".
  * Used to close the `HttpUrlConnection` when the stream has been closed.
- Added redirect endpoint fetching methods to Java "TiResponseCache".
  * Needed to fetch cached image when using a URL that redirects. Major optimization.

---
**Test 1:**
1. Uninstall the app from the device if already installed. (Will delete cached web responses.)
2. Make sure the Android device has Internet access.
3. Set "tiapp.xml" file property "run-on-main-thread" to `true`.
4. Build and run the below code.
5. Verify that the image was successfully displayed without crashing.

```javascript
var window = Ti.UI.createWindow();
window.add(Ti.UI.createView({
	// This URL redirects from HTTP to HTTPS.
	backgroundImage: "http://raw.githubusercontent.com/recurser/exif-orientation-examples/master/Portrait_1.jpg",
}));
window.open();
```

---
**Test 2:**
1. Uninstall the app from the device if already installed. (Will delete cached web responses.)
2. Make sure the Android device has Internet access.
3. Set "tiapp.xml" file property "run-on-main-thread" to `true`.
4. Build and run the below code.
5. Verify that the image was displayed "upright" successfully, without crashing.

```javascript
var window = Ti.UI.createWindow();
var imageView = Ti.UI.createImageView({
	// This URL redirects from HTTP to HTTPS.
	image: "http://raw.githubusercontent.com/recurser/exif-orientation-examples/master/Portrait_3.jpg",
	autorotate: true,
});
window.add(imageView);
window.open();
```
